pkg/operator/controllers/cloudproviderconfig/cloudproviderconfig_controller.go

```diff
@@ -133,31 +134,23 @@ type CloudProviderConfigReconciler struct {
 	base.AROController
 }
 
-func NewReconciler(Log *logrus.Entry, client client.Client) *CloudProviderConfigReconciler {
-	return &CloudProviderConfigReconciler{
+func NewReconciler(log *logrus.Entry, client client.Client) *CloudProviderConfigReconciler {
+	r := &CloudProviderConfigReconciler{
 		AROController: base.AROController{
-			Log:    Log,
-			Client: client,
-			Name:   ControllerName,
+			Log:         log.WithField("controller", controllerName),
+			Client:      client,
+			Name:        controllerName,
+			EnabledFlag: controllerEnabled,
 		},
 	}
+	r.Reconciler = r
```

Comment:
This recursive self-reference may indicate that a better type factoring exists. Is it possible to write a type that allows for composable components - which may share common references to shared input parameters - that is not self-referential?

Reply:
This is a way of emulating dynamic method dispatching in Go, since it's not native to the language.  (e.g. when `AROController` calls `c.Reconciler.ReconcileEnabled` we want to call `CloudProviderConfig`'s implementation)  But 20+ years of using object-oriented programming languages has me sometimes stuck in that mode of thinking.  Maybe there's a better way to organize in Go.

Reply:
Yes, I see the intent - it's very far from idiomatic, though. Composition of small atoms would feel much better.